### ISSUE TITLE
empty expansion of a macro pattern does not cause the comma-separated repetition expander to omit the comma

### ISSUE THREAD

**Issue by daym** [rust-lang/rust#96787]
I tried this code:

```rust
macro_rules! make_accessors {
(
        struct $struct_name:ident {
                $(
                        $(#[$field_meta:meta])*
                        $field_vis:vis $field_name:ident : $field_ty:ty
                        $(: get $getter_ty:ty)?,
                )* $(,)?
        }
) => {
        struct $struct_name {
                $(
                        $(#[$field_meta])*
                        $(
                                $field_name : $getter_ty
                        )?
                ),*
        }
}}

make_accessors! {
        struct A {
                a: u32 : get u32,
                b: u32,
                c: u32 : get u32,
        }
}

fn main() {
    println!("Hello, world!");
}
```

I expected to see this happen:

Since the expansion of the `$(...)?` and `$field_meta` is the empty expansion, I would have expected the `$(...),*` not to add a comma.

Instead, this happened:

It added a comma even though there was no expansion.

After that, the Rust compiler said:
```
error: expected identifier, found `,`
  --> src/main.rs:18:4
   |
18 |         ),*
   |          ^
   |          |
   |          expected identifier
   |          help: remove this comma
```

**Comment by Wardenfar** [comment]
Your macro declaration is incorrect.
In the pattern part of the macro, `$field_name` is at the 2nd level of "$ scopes", but in the replacement part of the macro :  `$field_name` is at the 3rd level of "$ scopes".

`$($field_name : $getter_ty)?`  
This part mix 2/3 level, and that's why you are maybe confused. 

When the compiler sees the field b without a `: get <type>` part.

`$field_name` has a value but not `$getter_ty`, so it gets replaced with nothing.
At the end, there is a blank after "b :" and rustc expect an identifier for the field type.

**Comment by daym** [comment]
>$field_name has a value but not $getter_ty, so it gets replaced with nothing.

Yes, I want the entire text `b: u32,`  to be replaced by nothing (in the case of a missing getter).

But if the $(...)? is not present, neither the $field_name nor the $getter_ty gets put, right?

You make it sound like `b:` still gets put in the case of $()? having 0 matches. Are you sure that's the case?

The goal is this:

I want Rust to create a struct with all the pub fields:

```
        struct A {
                a: u32,
                c: u32,
        }
```

but instead, it does this I think (I can't tell since `cargo expand` is also broken):

```
        struct A {
                a: u32 : get u32,
                ,
                c: u32 : get u32,
        }
```

**Comment by Wardenfar** [comment]
The reference says :

> Each repetition in the transcriber must contain at least one metavariable to decide how many times to expand it. If multiple metavariables appear in the same repetition, they must be bound to the same number of fragments. For instance, ( $( $i:ident ),* ; $( $j:ident ),* ) => (( $( ($i,$j) ),* )) must bind the same number of $i fragments as $j fragments. This means that invoking the macro with (a, b, c; d, e, f) is legal and expands to ((a,d), (b,e), (c,f)), but (a, b, c; d, e) is illegal because it does not have the same number. This requirement applies to every layer of nested repetitions.

And if I understand correctly : one of them is present, so the scope is not ignored. (repeated 1 time)

[Rust Reference : Macros Repetitions](https://doc.rust-lang.org/stable/reference/macros-by-example.html#repetitions.). 

After this, i can't help much more :)

**Comment by Wardenfar** [comment]
Nevermind, i made it work by moving the comma in the $ scope.
You were right.

```
macro_rules! make_accessors {
    (
            struct $struct_name:ident {
                    $(
                            $(#[$field_meta:meta])*
                            $field_vis:vis $field_name:ident : $field_ty:ty
                            $(: get $getter_ty:ty)?,
                    )* $(,)?
            }
    ) => {
            struct $struct_name {
                    $(
                            $(#[$field_meta])*
                            $(
                                    $field_name : $getter_ty, // put it there
                            )?
                    )* // remove the comma here
            }
    }}
    make_accessors! {
            struct A {
                    a: u32 : get u32,
                    b: u32,
                    c: u32 : get u32,
            }
    }
    
    fn main() {
    }
```

**Comment by daym** [comment]
The problem with that is that if there's a field with meta and without get, then it will still output the meta and then that will apply to the next field.

For example

```
    make_accessors! {
            struct A {
                    #[quux]
                    a: u32 : get Q,
                    #[foo]
                    b: u32,
                    #[bar]
                    c: u32 : get R,
            }
    }
```

With your variant, the following gets output:

```
            struct A {
                    #[quux]
                    a: Q,
                    #[foo] // uh oh
                    #[bar]
                    c: R,
            }
```

This is very very bad for my purposes.

**Comment by Wardenfar** [comment]
Ok, I don't know how to solve this problem with declarative macros.
You should consider using proc-macros. #[derive(****)]
So this is not a bug. Maybe close this issue ?

**Comment by daym** [comment]
I think emitting an extra comma between nothing is a bug.

Doing so doesn't seem like something `$(...),*` should be doing--after all, that's why there is a special comma form in the first place.

**Comment by Wardenfar** [comment]
> Also, cargo expand erroring out and just skipping the entire thing is a bug. I mean what are you using cargo expand for if not to debug broken macro replacements? And in such a case it just skips the output entirely?

`cargo expand` show the result after expansion, but in this case the expansion failed, `cargo expand` cannot print anything.

**Comment by daym** [comment]
> `cargo expand` show the result after expansion, but in this case the expansion failed, `cargo expand` cannot print anything.

What do you use in such a case to find what it did?
